cmd/tf_operator/main.go

```diff
@@ -83,6 +86,23 @@ func init() {
 		log.Info("No controller_config_file provided; using empty config.")
 	}
 
+	if err := checkGRPCServerPath(); err != nil {
+		log.Errorf("Errored when checkGRPCServerPath: %v", err)
+		os.Exit(1)
+	}
+}
+
+// checkGRPCServerPath checks if GrpcServerFilePath is defined in controller_config_file or CLI argument, if not return an error
+func checkGRPCServerPath() error {
+	if gRPCServerFilePath != "" {
```

Comment:
Why do we need two ways of specifying the gRPCServerFilePath?

Reply:
After thinking about this more it should probably just be a flag and not specified via the config file.

The config file is intended to be overwritten by cluster admins to customize the config for their cluster.

But gRPCServerFilePath isn't really intended for users to customize. If users want to customize the server (e.g. by specifying their own program) they should do this just by specifying the container image and command to use for the PS replica.

The default gRPC server program is baked into the container for the controller. Ideally, the location of the gRPC program would just be hard coded. But that would be brittle because the location is specified in the Docker file.  It would also be problematic if the user is running the controller in debug mode locally and not in a container. Allowing the path to be specified by a flag solves these problems.

The problem with specifying it in the config file, is that if a user then wants to override some of the config they end up having to specify the gRPCServerFilePath which isn't good.

Reply:
It makes sense, I will reimplement it.